lib/tpm/tpm.go

```diff
@@ -0,0 +1,216 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tpm
+
+import (
+	"context"
+	"crypto"
+	"crypto/sha256"
+	"crypto/x509"
+	"fmt"
+	"log/slog"
+	"math/big"
+	"strings"
+
+	"github.com/google/go-attestation/attest"
+	"github.com/gravitational/trace"
+	"go.opentelemetry.io/otel"
+)
+
+var tracer = otel.Tracer("github.com/gravitational/teleport/lib/tpm")
+
+// serialString converts a serial number into a readable colon-delimited hex
+// string thats user-readable e.g ab:ab:ab:ff:ff:ff
+func serialString(serial *big.Int) string {
+	hex := serial.Text(16)
+	if len(hex)%2 == 1 {
+		hex = "0" + hex
+	}
+
+	out := strings.Builder{}
+	for i := 0; i < len(hex); i += 2 {
+		if i != 0 {
+			out.WriteString(":")
+		}
+		out.WriteString(hex[i : i+2])
+	}
+	return out.String()
+}
+
+// hashEKPub hashes the public part of an EK key. The key is first marshaled
+// in PKIX format and then hashed with SHA256.
+func hashEKPub(key crypto.PublicKey) (string, error) {
+	marshaled, err := x509.MarshalPKIXPublicKey(key)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	hashed := sha256.Sum256(marshaled)
+	return fmt.Sprintf("%x", hashed), nil
+}
+
+// QueryRes is the result of the TPM query performed by Query.
+type QueryRes struct {
+	// EKPub is the PKIX marshaled public part of the EK.
+	EKPub []byte
+	// EKPubHash is the SHA256 hash of the PKIX marshaled EKPub.
+	EKPubHash string
+	// EKCertPresent is true if an EK cert is present in the TPM.
+	EKCertPresent bool
+	// EKCert is the ASN.1 DER encoded EK cert.
+	EKCert []byte
+	// EKCertSerial is the serial number of the EK cert represented as a colon
+	// delimited hex string.
+	EKCertSerial string
+}
+
+// Query returns information about the TPM on a system, including the
+// EKPubHash and EKCertSerial which are needed to configure TPM joining.
+func Query(ctx context.Context, log *slog.Logger) (*QueryRes, error) {
```

Comment:
🤔  tough one - I'm very much in the camp of treating loggers like all other dependencies and injecting them rather than using global loggers. It makes it way easier to add fields which should be included in all logs emitted by child functions. Another thought is I'd love to fix how badly garbled our parallel test log output is one day, and the only way to do that is to avoid global loggers and to be able to inject a logger which properly attributes logs to the test they belong to.

Reply:
The peeve I have with this is that it "messes up" the function signature. Taking to an extreme then almost every standalone func suddenly takes a logger.

I suppose this is a reflection of us not having a good way to "inject" orthogonal concerns like loggers. A slightly better way to solve this is tying the methods to a struct and pushing the logger to it.

Feel free to resolve this thread.

Reply:
> almost every standalone func suddenly takes a logger.

Perhaps I'm a bit of an extremist here - but I do unironically believe that this is how things should be done, logger injected as a parameter or as a field struct. In my experience with Go so far, Globals have repeatedly caused large amounts of pain, whereas DI is more of a minor annoyance 😓